File: top_interview_questions/1_easy/6_dynamic_programming/best_time_to_buy_and_sell_stock/solution.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0

        max_val = min_val = prices[0]
        max_gain = current_gain = 0

        for index in range(1, len(prices)):

            if prices[index] > max_val:
                max_val = prices[index]
                current_gain = max_val - min_val
                max_gain = current_gain if max_gain < current_gain else max_gain

            if prices[index] < min_val:
                min_val = max_val = prices[index]
                current_gain = 0

        return max_gain

File: top_interview_questions/1_easy/6_dynamic_programming/best_time_to_buy_and_sell_stock/solution.go
package main

func main() {

}

func maxProfit(prices []int) int {
	if len(prices) == 0 {
		return 0
	}

	maxVal, minVal := prices[0], prices[0]
	maxGain, currGain := 0, 0

	for _, price := range prices[1:] {
		if price > maxVal {
			maxVal = price
			currGain = maxVal - minVal
			if currGain > maxGain {
				maxGain = currGain
			}
		}

		if price < minVal {
			minVal = price
			maxVal = price
			currGain = 0
		}
	}

	return maxGain
}
